### DISCUSSION TITLE
Getting the selected value from a Plot

### DISCUSSION THREAD

**Discussion by Kerollmops** [emilk/egui#1778] [Q&A]
Hey,

I would like to know if there is a way to get the value that has been auto-snap-selected by egui when the cursor moves nearby.
I know that we can get [the exact position of the cursor by using `PlotUi::pointer_coordinate`](https://docs.rs/egui/latest/egui/widgets/plot/struct.PlotUi.html#method.pointer_coordinate) but it is not exactly what I would like.

For example on this screenshot, my cursor is at something like 30px from the auto-snap-selected value.

<img width="173" alt="Capture d’écran 2022-06-25 à 19 20 55" src="https://user-images.githubusercontent.com/3610253/176547434-e54e1e00-5c64-4b45-a764-c5690c4717c5.png">

**Comment by benedikt-jensen** [comment]
I have the same problem.

**Comment by Peepo-Juice** [comment]
I'm working on something right now and would like this feature as well.

**Comment by haricot** [comment]
I propose a pull request #2572 that covers this aspect, it not directly retrieve the pointer coordinate selected by the radius but give if is_in_radius, hovered indexes and item name hovered  after that you can retrieve information about the selected value from a item Plot.

**Comment by Turun** [comment]
You can use [`Plot::label_formatter()`](https://docs.rs/egui_plot/latest/egui_plot/struct.Plot.html#method.label_formatter) as a workaround. This method is called in order to determine what should be written in the little popup and therefore contains the coordinates of the point you are interested in. 

The obvious implementation of this workaround - cloning the `PlotPoint` parameter given in the `label_formatter` closure - doesn't compile, however. The rust compile can not guarantee that the closure will live shorter than the parameter you want to copy it to. So you need to pass a static variable or something else owned into the closure that can report back the value of the `PlotPoint` parameter whenever the closure is run (as far as the compiler is concerned it may be stored and executed much later, after the current call of `App::update` already terminated and your extraction variable is dropped).   
I used a channel for that. It's very hacky, but it works. It also allows me to check if the crosshair is not painted in the plot this frame. Very important if you can not afford to work with outdated information about the crosshair from a previous frame.

My solution looks like this:

```rust
let (tx, rx) = mpsc::channel::<PlotPoint>();
let plot_response = Plot::new("known implants plot")
    .label_formatter(move |_name, value| {
        let _response = tx.send(value.clone());
        format!("x={value.x}\ny={value.y}")
    })
    .show(ui, |ui| { ui.points(points) } );

let crosshair: PlotPoint = if let Ok(value) = rx.recv() {
    Some(value)
} else {
    // nothing was sent, e.g. because the user doesn't hover the plot
    None
};
```

Additionally, you can check if the crosshair is snapped to a plot item or not by doing:

```rust
let mouse_position = plot_response.response.hover_pos();
if let Some(crosshair) = crosshair {
    // check if the position of the mouse and the position of the crosshair
    // are identical (in screenspace here, but plotspace should work too)
    let crosshair_is_snapped = mouse_position != plot_response.transform.position_from_point(&crosshair);
}
```
